### DISCUSSION TITLE
Upgrading to 3.0 & IDE pollution -> Unable to acquire the logger context

### DISCUSSION THREAD

**Discussion by natan-abolafya** [dropwizard/dropwizard#7108] [Q&A] [answered]
I'm in the middle of upgrading dropwizard from 2.1.6 to 3.0 and struggling to start the application (or run tests) using Intellij due to this error
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Exception in thread "main" java.lang.IllegalStateException: Unable to acquire the logger context
	at io.dropwizard.logging.common.LoggingUtil.getLoggerContext(LoggingUtil.java:46)
	at io.dropwizard.logging.common.BootstrapLogging.bootstrap(BootstrapLogging.java:51)
	at io.dropwizard.logging.common.BootstrapLogging.bootstrap(BootstrapLogging.java:40)
	at io.dropwizard.core.Application.bootstrapLogging(Application.java:38)
	at io.dropwizard.core.Application.<init>(Application.java:26)
	at com.appgate.controller.dropwizard.application.BaseRestApplication.<init>(BaseRestApplication.java:53)
	at com.appgate.controller.ControllerApplication.<init>(ControllerApplication.java:62)
```

It works if I run it using gradle instead of intellij, so it's gotta mean IDE pollution. But I can't seem to unpollute the IDE and I find working with IDE more efficient so I don't want to give up on that yet. I thought I'd ask help here on where to look.

It's a relatively big project with a lot of dependencies, so it's a bit hard to give details without dumping lots of information here. Here is what I have been trying though:
- I have been removing test libraries and excluding transitive dependencies from other libraries which I think might interfere.
- Removed all libraries from intellij and reload gradle.
- Added `slf4j-simple-2.0.7` as dependency.
- Looking at both gradle.lock file and the "External Libraries" Intellij reports but nothing looks out of place.
![image](https://github.com/dropwizard/dropwizard/assets/64091305/e7a8b220-217e-4c38-b183-faa39ed482c6)
![image](https://github.com/dropwizard/dropwizard/assets/64091305/8ce217c0-0200-44be-a6c7-9a7f0950f458)

Any other ideas?

**Comment by joschi** [comment]
From https://www.slf4j.org/codes.html#StaticLoggerBinder

> Note that slf4j-api versions 2.0.x and later use the [ServiceLoader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html) mechanism. Backends such as logback 1.3 and later which target slf4j-api 2.x, do not ship with org.slf4j.impl.StaticLoggerBinder. If you place a logging backend which targets slf4j-api 2.0.x, you need slf4j-api-2.x.jar on the classpath. See also [relevant faq](https://www.slf4j.org/faq.html#changesInVersion18) entry.

Are you 100% sure that only SLF4J 2.0.7 is on the classpath and that no older version is left over in IntelliJ IDEA?

**Reply by natan-abolafya** [reply]
According to the screenshots I've pasted, it looks like the right version. Do you know any other way to look at it?

By the way, it might not be IDE pollution. If I run the application with gradle it works. but when I build the application with gradle, I come across the same issue. And I can see that the `slf4j-api-2.0.7.jar` is in the dependency directory.

I've tried to reproduce it with dropwizard-example project (maven) and adding pretty much all the dependencies we have and I could not manage to reproduce the issue.

**Reply by natan-abolafya** [reply, marked as answer]
Having read [this] https://github.com/dropwizard/dropwizard/discussions/6931#discussioncomment-6024492) like the third time after a long break, I finally understood what the problem was. There was a dependency that had the package `org.slf4j` included as code (I don't remember what this practice is called now). Found a version of the library that does transitive dependencies instead.

Thanks for the help.

**Reply by joschi** [reply]
Thanks for sharing how you resolved the issue! 😄

**Reply by natan-abolafya** [reply]
:) no problem. I should also add how I found the culprit library in case it helps someone else.
I went into slf4j library, selected a class (like LoggerFactory but it was a bit too generic tbh) and then searched for classes (ctrl-n) with the same name in Intellij where libraries are included like so
![image](https://github.com/dropwizard/dropwizard/assets/64091305/a8947ed8-ef85-4991-bf55-c1caf6709d6f)
and I saw found one with package `org.slf4j` but on another library.
